File: Data Structures & Algorithms/top-k-elements-in-list/submission-3.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        vector<vector<int>> buckets(nums.size()+1);
        vector<int> res;

        for (int num : nums){
            freq[num]++;
        }

        for (const auto& [num,count] : freq){
            buckets[count].push_back(num);
        }

        for (int ct = nums.size(); ct > 0 && res.size() < k; --ct){
            for (int num : buckets[ct]){
                res.push_back(num);
                if (res.size() == k) return res;
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/kth-largest-element-in-an-array/submission-3.cs
public class Solution {
    public int FindKthLargest(int[] nums, int k) {
        var minHeap = new PriorityQueue<int,int>();

        foreach (int num in nums){
            if (minHeap.Count < k){
                minHeap.Enqueue(num,num);
            }
            else if (num > minHeap.Peek()){
                minHeap.Dequeue();
                minHeap.Enqueue(num,num);
            }
        }

        return minHeap.Peek();
    }
}
